batch/sql/improve-cancel.sql

```diff
@@ -0,0 +1,561 @@
+
+ALTER TABLE `batches` ADD COLUMN `state` VARCHAR(40);
+
+UPDATE `batches`
+SET `state` = CASE
+  WHEN NOT closed THEN 'open'
+  WHEN n_completed = n_jobs THEN 'complete'
+  ELSE 'running'
+  END;
+
+ALTER TABLE `batches`
+  MODIFY COLUMN `state` VARCHAR(40) NOT NULL,
+  DROP COLUMN closed;
+
+DROP INDEX `batches_user` ON `batches`;
+CREATE INDEX `batches_user_state_cancelled` ON `batches` (`user`, `state`, `cancelled`);
+
+DROP INDEX `jobs_state` ON `jobs`;
+CREATE INDEX `jobs_batch_id_state_always_run_cancelled` ON `jobs` (`batch_id`, `state`, `always_run`, `cancelled`);
+
+ALTER TABLE `user_resources`
+  ADD COLUMN `n_cancelled_ready_jobs` INT NOT NULL DEFAULT 0,
+  ADD COLUMN `n_cancelled_running_jobs` INT NOT NULL DEFAULT 0;
+
+CREATE TABLE `batch_cancellable_resources` (
+  `batch_id` BIGINT NOT NULL,
+  `token` INT NOT NULL,
+  # neither run_always nor cancelled
+  `n_ready_cancellable_jobs` INT NOT NULL DEFAULT 0,
+  `ready_cancellable_cores_mcpu` BIGINT NOT NULL DEFAULT 0,
+  `n_running_cancellable_jobs` INT NOT NULL DEFAULT 0,
+  `running_cancellable_cores_mcpu` BIGINT NOT NULL DEFAULT 0,
+  PRIMARY KEY (`batch_id`, `token`),
+  FOREIGN KEY (`batch_id`) REFERENCES batches(id) ON DELETE CASCADE
+) ENGINE = InnoDB;
+
+DELIMITER $$
+
+CREATE PROCEDURE recompute_incremental(
+) BEGIN
+
+  DELETE FROM batches_staging;
+  DELETE FROM batch_cancellable_resources;
+  DELETE FROM ready_cores;
+  DELETE FROM user_resources;
+
+  DROP TEMPORARY TABLE IF EXISTS `tmp_batch_resources`;
+
+  CREATE TEMPORARY TABLE `tmp_batch_resources` AS (
+    SELECT batch_id, batch_state, user,
+      COALESCE(SUM(1), 0) as n_jobs,
+      COALESCE(SUM(job_state = 'Ready' AND cancellable), 0) as n_ready_cancellable_jobs,
+      COALESCE(SUM(IF(job_state = 'Ready' AND cancellable, cores_mcpu, 0)), 0) as ready_cancellable_cores_mcpu,
+      COALESCE(SUM(job_state = 'Running' AND NOT cancelled), 0) as n_running_jobs,
+      COALESCE(SUM(IF(job_state = 'Running' AND NOT cancelled, cores_mcpu, 0)), 0) as running_cores_mcpu,
```

Comment:
This changes the definition of running jobs from what it was before. I think that's okay because it's used in user_resources to compute the fair share if I remember correctly. However, this value doesn't represent the number of cores actually occupied in the system.

Reply:
Yes.  That's one of the main points of this change.  What we want for scaling the instance pool and calculating shares is not ready jobs, but runnable jobs.  That's why I wanted to rename ready => runnable.  I will still probably do this in a later PR for clarity.